File: articles/calculus-for-systems-modeling/articles/constrained-optimization-and-lagrange-multipliers/julia/constrained_optimization_audit.jl
objective(x, y) = x^2 + 2*y^2
constraint(x, y) = x + y
grad_objective(x, y) = (2*x, 4*y)
grad_constraint(x, y) = (1.0, 1.0)

function solve_budget_constraint(target)
    y = target / 3
    x = 2 * target / 3
    lambda_value = 2 * x
    return x, y, lambda_value
end

println("x,y,objective_value,constraint_value,constraint_target,constraint_residual,lambda_value,gradient_f_x,gradient_f_y,gradient_g_x,gradient_g_y,stationarity_residual_norm,feasible,warning")
for target in [12.0, 18.0, 24.0]
    x, y, lambda_value = solve_budget_constraint(target)
    gfx, gfy = grad_objective(x, y)
    ggx, ggy = grad_constraint(x, y)
    stationarity_x = gfx - lambda_value * ggx
    stationarity_y = gfy - lambda_value * ggy
    stationarity_residual_norm = sqrt(stationarity_x^2 + stationarity_y^2)
    cval = constraint(x, y)
    cres = cval - target
    feasible = abs(cres) <= 1e-9
    warning = feasible && stationarity_residual_norm <= 1e-8 ? "Multiplier interpretation is local and unit-dependent." : "Review feasibility or stationarity."
    println("$x,$y,$(objective(x,y)),$cval,$target,$cres,$lambda_value,$gfx,$gfy,$ggx,$ggy,$stationarity_residual_norm,$feasible,$warning")
end

File: articles/calculus-for-systems-modeling/articles/constrained-optimization-and-lagrange-multipliers/c/constrained_optimization_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double objective(double x, double y){ return x*x + 2.0*y*y; }

int main(void){
  double targets[3] = {12.0, 18.0, 24.0};
  printf("x,y,objective_value,constraint_value,constraint_target,constraint_residual,lambda_value,gradient_f_x,gradient_f_y,gradient_g_x,gradient_g_y,stationarity_residual_norm,feasible,warning\n");
  for(int i=0; i<3; i++){
    double target = targets[i];
    double y = target / 3.0;
    double x = 2.0 * target / 3.0;
    double lambda = 2.0 * x;
    double gfx = 2.0*x, gfy = 4.0*y, ggx = 1.0, ggy = 1.0;
    double sx = gfx - lambda*ggx, sy = gfy - lambda*ggy;
    double norm = sqrt(sx*sx + sy*sy);
    double cval = x + y;
    double cres = cval - target;
    int feasible = fabs(cres) <= 1e-9;
    const char* warning = feasible && norm <= 1e-8 ? "Multiplier interpretation is local and unit-dependent." : "Review feasibility or stationarity.";
    printf("%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%.12f,%d,%s\n", x,y,objective(x,y),cval,target,cres,lambda,gfx,gfy,ggx,ggy,norm,feasible,warning);
  }
  return EXIT_SUCCESS;
}
